File: src/android/ad_mob/src/main/java/com/ee/internal/AdMobNativeAd.kt
package com.ee.internal

import android.app.Activity
import android.content.Context
import android.graphics.Point
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.widget.Button
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.RatingBar
import android.widget.TextView
import androidx.annotation.AnyThread
import androidx.annotation.UiThread
import com.ee.IBannerAd
import com.ee.ILogger
import com.ee.IMessageBridge
import com.ee.Thread
import com.ee.Utils
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdLoader
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.VideoController.VideoLifecycleCallbacks
import com.google.android.gms.ads.formats.MediaView
import com.google.android.gms.ads.formats.UnifiedNativeAd
import com.google.android.gms.ads.formats.UnifiedNativeAdView
import kotlinx.serialization.InternalSerializationApi
import kotlinx.serialization.Serializable
import java.util.concurrent.atomic.AtomicBoolean

private typealias ViewProcessor<T> = (view: T) -> Unit

/**
 * Created by Zinge on 10/16/17.
 */
@InternalSerializationApi
internal class AdMobNativeAd(
    private val _bridge: IMessageBridge,
    private val _logger: ILogger,
    private val _context: Context,
    private var _activity: Activity?,
    private val _adId: String,
    private val _layoutName: String,
    private val _identifiers: Map<String, String>)
    : IBannerAd, UnifiedNativeAd.OnUnifiedNativeAdLoadedListener, AdListener() {
    @Serializable
    @Suppress("unused")
    private class ErrorResponse(
        val code: Int,
        val message: String
    )
    
    companion object {
        private val kTag = AdMobNativeAd::class.java.name
        private const val k__body = "body"
        private const val k__call_to_action = "call_to_action"
        private const val k__headline = "headline"
        private const val k__icon = "icon"
        private const val k__image = "image"
        private const val k__media = "media"
        private const val k__price = "price"
        private const val k__star_rating = "star_rating"
        private const val k__store = "store"
    }

    private val _messageHelper = MessageHelper("AdMobNativeAd", _adId)
    private val _helper = BannerAdHelper(_bridge, this, _messageHelper)
    private val _viewHelper = ViewHelper(Point(0, 0), Point(0, 0), false)
    private val _isLoaded = AtomicBoolean(false)
    private var _ad: AdLoader? = null
    private var _view: UnifiedNativeAdView? = null

    init {
        _logger.info("$kTag: constructor: adId = $_adId")
        registerHandlers()
        createInternalAd()
        createView()
    }

    override fun onCreate(activity: Activity) {
        _activity = activity
        addToActivity()
    }

    override fun onResume() {
    }

    override fun onPause() {
    }

    override fun onDestroy() {
        removeFromActivity()
        _activity = null
    }

    override fun destroy() {
        _logger.info("$kTag: ${this::destroy.name}: adId = $_adId")
        deregisterHandlers()
        destroyView()
        destroyInternalAd()
    }

    @AnyThread
    private fun registerHandlers() {
        _helper.registerHandlers()
    }

    @AnyThread
    private fun deregisterHandlers() {
        _helper.deregisterHandlers()
    }

    @AnyThread
    private fun createInternalAd() {
        Thread.runOnMainThread {
            if (_ad != null) {
                return@runOnMainThread
            }
            _isLoaded.set(false)
            _ad = AdLoader.Builder(_context, _adId)
                .forUnifiedNativeAd(this)
                .withAdListener(this)
                .build()
        }
    }

    @AnyThread
    private fun destroyInternalAd() {
        Thread.runOnMainThread {
            if (_ad == null) {
                return@runOnMainThread
            }
            _isLoaded.set(false)
            _ad = null
        }
    }

    @AnyThread
    private fun createView() {
        Thread.runOnMainThread {
            val layoutId = _context.resources
                .getIdentifier(_layoutName, "layout", _context.packageName)
            val view = LayoutInflater
                .from(_context)
                .inflate(layoutId, null, false)
            view.visibility = View.INVISIBLE
            val params = FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.WRAP_CONTENT,
                FrameLayout.LayoutParams.WRAP_CONTENT)
            params.gravity = Gravity.START or Gravity.TOP
            view.layoutParams = params
            _view = view as UnifiedNativeAdView
            _viewHelper.view = view
            addToActivity()
        }
    }

    @AnyThread
    private fun destroyView() {
        Thread.runOnMainThread {
            removeFromActivity()
            _view = null
            _viewHelper.view = null
        }
    }

    @AnyThread
    private fun addToActivity() {
        Thread.runOnMainThread {
            val activity = _activity ?: return@runOnMainThread
            val rootView = Utils.getRootView(activity)
            rootView.addView(_view)
        }
    }

    @UiThread
    private fun removeFromActivity() {
        Thread.runOnMainThread {
            val activity = _activity ?: return@runOnMainThread
            val rootView = Utils.getRootView(activity)
            rootView.removeView(_view)
        }
    }

    override val isLoaded: Boolean
        @AnyThread get() = _isLoaded.get()

    @AnyThread
    override fun load() {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::load.name}: id = $_adId")
            val ad = _ad ?: throw IllegalArgumentException("Ad is not initialized")
            ad.loadAd(AdRequest.Builder().build())
        }
    }

    override var position: Point
        @AnyThread get() = _viewHelper.position
        @AnyThread set(value) {
            _viewHelper.position = value
        }

    override var size: Point
        @AnyThread get() = _viewHelper.size
        @AnyThread set(value) {
            _viewHelper.size = value
        }

    override var isVisible: Boolean
        @AnyThread get() = _viewHelper.isVisible
        @AnyThread set(value) {
            _viewHelper.isVisible = value
        }

    private fun getIdentifier(identifier: String): Int {
        if (!_identifiers.containsKey(identifier)) {
            return 0
        }
        val resources = _context.resources
        return resources.getIdentifier(_identifiers[identifier], "id", _context.packageName)
    }

    @UiThread
    private fun <T : View> processView(view: View,
                                       key: String,
                                       processor: ViewProcessor<T>): Boolean {
        val id = getIdentifier(key)
        if (id == 0) {
            _logger.error("$kTag: Can not find identifier for key: $key")
            return false
        }
        val subView = view.findViewById<View>(id)
        if (subView == null) {
            _logger.error("$kTag: Can not find view for key: $key")
            return false
        }
        processor(subView as T)
        return true
    }

    override fun onUnifiedNativeAdLoaded(ad: UnifiedNativeAd) {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::onUnifiedNativeAdLoaded.name}")
            val view = _view ?: throw IllegalArgumentException("Ad is not initialized")

            // Get the video controller for the ad. One will always be provided, even if the ad doesn't
            // have a video asset.
            val vc = ad.videoController

            // Create a new VideoLifecycleCallbacks object and pass it to the VideoController. The
            // VideoController will call methods on this object when events occur in the video
            // lifecycle.
            vc.videoLifecycleCallbacks = object : VideoLifecycleCallbacks() {
                override fun onVideoEnd() {
                    // Publishers should allow native ads to complete video playback before refreshing
                    // or replacing them with another ad in the same UI location.
                    _logger.debug("$kTag: ${this::onVideoEnd.name}")
                    super.onVideoEnd()
                }
            }

            // Some assets are guaranteed to be in every NativeAppInstallAd.
            processView<TextView>(view, k__body) { item ->
                view.bodyView = item
                item.text = ad.body
            }
            processView<Button>(view, k__call_to_action) { item ->
                view.callToActionView = item
                item.text = ad.callToAction
            }
            processView<TextView>(view, k__headline) { item ->
                view.headlineView = item
                item.text = ad.headline
            }
            processView<ImageView>(view, k__icon) { item ->
                view.iconView = item
                if (ad.icon != null) {
                    item.setImageDrawable(ad.icon.drawable)
                }
            }

            // Apps can check the VideoController's hasVideoContent property to determine if the
            // NativeAppInstallAd has a video asset.
            if (vc.hasVideoContent()) {
                processView<ImageView>(view, k__image) { item ->
                    item.visibility = View.GONE
                }
                processView<MediaView>(view, k__media) { item ->
                    view.mediaView = item
                }
            } else {
                processView<ImageView>(view, k__image) { item ->
                    val images = ad.images
                    item.visibility = View.GONE
                    for (image in images) {
                        if (image != null) {
                            item.visibility = View.VISIBLE
                            item.setImageDrawable(image.drawable)
                            break
                        }
                    }
                }
                processView<MediaView>(view, k__media) { item ->
                    item.visibility = View.GONE
                }
            }

            // These assets aren't guaranteed to be in every NativeAppInstallAd, so it's important to
            // check before trying to display them.
            processView<TextView>(view, k__price) { item ->
                view.priceView = item
                val price: CharSequence? = ad.price
                if (price != null) {
                    item.visibility = View.VISIBLE
                    item.text = price
                } else {
                    item.visibility = View.INVISIBLE
                }
            }
            processView<RatingBar>(view, k__star_rating) { item ->
                view.starRatingView = item
                val starRating = ad.starRating
                if (starRating != null) {
                    item.rating = starRating.toFloat()
                    item.visibility = View.VISIBLE
                } else {
                    item.visibility = View.INVISIBLE
                }
            }
            processView<TextView>(view, k__store) { item ->
                view.storeView = item
                val store: CharSequence? = ad.store
                if (store != null) {
                    item.visibility = View.VISIBLE
                    item.text = store
                } else {
                    item.visibility = View.INVISIBLE
                }
            }

            // Assign native ad object to the native view.
            view.setNativeAd(ad)

            _isLoaded.set(true)
            _bridge.callCpp(_messageHelper.onLoaded)
        }
    }

    override fun onAdLoaded() {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::onAdLoaded.name}: id = $_adId")
            // There is one important difference between the way AdListener objects work with native ads
            // and the way they work with banners and interstitials. Because the AdLoader has its own
            // format-specific listeners (i.e., UnifiedNativeAd.OnUnifiedNativeAdLoadedListener) to use
            // when an ad has loaded, the onAdLoaded() method from AdListener is not called when a
            // native ad loads successfully.
        }
    }

    override fun onAdFailedToLoad(error: LoadAdError) {
        Thread.runOnMainThread {
            _logger.debug("$kTag: onAdFailedToLoad: id = $_adId message = ${error.message}")
            _bridge.callCpp(_messageHelper.onFailedToLoad, ErrorResponse(error.code, error.message).serialize())
        }
    }

    override fun onAdOpened() {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::onAdOpened.name}: id = $_adId")
        }
    }

    override fun onAdImpression() {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::onAdImpression.name}: id = $_adId")
        }
    }

    override fun onAdClicked() {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::onAdClicked.name}: id = $_adId")
            _bridge.callCpp(_messageHelper.onClicked)
        }
    }

    override fun onAdClosed() {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::onAdClosed.name}: id = $_adId")
        }
    }
}

File: src/ios/ee/ad_mob/AdMobNativeAd.swift
//
//  AdMobNativeAd.swift
//  ee-x-366f64ba
//
//  Created by eps on 6/24/20.
//

import Foundation
import GoogleMobileAds

/*
  FIXME.
 
private let kTag = "\(AdMobNativeAd.self)"

/// https://stackoverflow.com/a/24592029
func dictionaryOfNames(_ arr: UIView...) -> [String: UIView] {
    var d = [String: UIView]()
    for (ix, v) in arr.enumerated() {
        d["v\(ix + 1)"] = v
    }
    return d
}

internal class AdMobNativeAd: NSObject, IBannerAd,
    GADUnifiedNativeAdLoaderDelegate,
    GADUnifiedNativeAdDelegate {
    private let _bridge: IMessageBridge
    private let _logger: ILogger
    private let _adId: String
    private let _layoutName: String
    private let _messageHelper: MessageHelper
    private var _helper: BannerAdHelper?
    private let _viewHelper: ViewHelper
    private var _isLoaded = false
    private var _ad: GADAdLoader?
    private var _view: GADUnifiedNativeAdView?

    init(_ bridge: IMessageBridge,
         _ logger: ILogger,
         _ adId: String,
         _ layoutName: String) {
        _bridge = bridge
        _logger = logger
        _adId = adId
        _layoutName = layoutName
        _messageHelper = MessageHelper("AdMobNativeAd", _adId)
        _viewHelper = ViewHelper(CGPoint.zero, CGSize.zero, false)
        super.init()
        _helper = BannerAdHelper(_bridge, self, _messageHelper)
        registerHandlers()
        createInternalAd()
        createView()
    }

    func destroy() {
        deregisterHandlers()
        destroyView()
        destroyInternalAd()
    }

    func registerHandlers() {
        _helper?.registerHandlers()
    }

    func deregisterHandlers() {
        _helper?.deregisterHandlers()
    }

    func createInternalAd() {
        Thread.runOnMainThread {
            if self._ad != nil {
                return
            }
            self._isLoaded = false
            let options = GADVideoOptions()
            options.startMuted = true
            let rootView = Utils.getCurrentRootViewController()
            let ad = GADAdLoader(adUnitID: self._adId,
                                 rootViewController: rootView,
                                 adTypes: [GADAdLoaderAdType.unifiedNative],
                                 options: [options])
            ad.delegate = self
            self._ad = ad
        }
    }

    func destroyInternalAd() {
        Thread.runOnMainThread {
            guard let ad = self._ad else {
                return
            }
            self._isLoaded = false
            ad.delegate = nil
            self._ad = nil
        }
    }

    func createAdView() -> GADUnifiedNativeAdView? {
        guard let nibObjects = Bundle(for: GADUnifiedNativeAdView.self).loadNibNamed(_layoutName,
                                                                                     owner: nil,
                                                                                     options: nil) else {
            assert(false, "Invalid layout")
            return nil
        }
        guard let view = nibObjects[0] as? GADUnifiedNativeAdView else {
            assert(false, "Invalid layout class")
            return nil
        }
        return view
    }

    func createView() {
        Thread.runOnMainThread {
            assert(self._view == nil)
            guard let view = self.createAdView() else {
                assert(false, "Cannot create ad view")
                return
            }
            self._view = view
            self._viewHelper.view = view
            let rootView = Utils.getCurrentRootViewController()
            rootView?.view.addSubview(view)
        }
    }

    func destroyView() {
        Thread.runOnMainThread {
            guard let view = self._view else {
                return
            }
            view.removeFromSuperview()
            self._view = nil
            self._viewHelper.view = nil
        }
    }

    var isLoaded: Bool {
        return _isLoaded
    }

    func load() {
        Thread.runOnMainThread {
            guard let ad = self._ad else {
                assert(false, "Ad is not initialized")
                return
            }
            ad.load(GADRequest())
        }
    }

    var position: CGPoint {
        get { return _viewHelper.position }
        set(value) { _viewHelper.position = value }
    }

    var size: CGSize {
        get { return _viewHelper.size }
        set(value) { _viewHelper.size = value }
    }

    var isVisible: Bool {
        get { return _viewHelper.isVisible }
        set(value) {
            _viewHelper.isVisible = value
            if value {
                Thread.runOnMainThread {
                    self._view?.subviews.forEach { $0.setNeedsDisplay() }
                }
            }
        }
    }

    /// Gets an image representing the number of stars. Returns nil if rating is
    /// less than 3.5 stars.
    func imageForStars(_ stars: Float) -> UIImage? {
        switch stars {
            case _ where stars >= 5: return UIImage(named: "stars_5")
            case _ where stars >= 4.5: return UIImage(named: "stars_4_5")
            case _ where stars >= 4: return UIImage(named: "stars_4")
            case _ where stars >= 3.5: return UIImage(named: "stars_3_5")
            default: return nil
        }
    }

    func adLoader(_ adLoader: GADAdLoader, didReceive ad: GADUnifiedNativeAd) {
        Thread.runOnMainThread {
            guard let view = self._view else {
                assert(false, "Ad view is not initialized")
                return
            }

            // Populate the native ad view with the native ad assets.
            // The headline and mediaContent are guaranteed to be present in every
            // native ad.
            (view.headlineView as? UILabel)?.text = ad.headline
            view.mediaView?.mediaContent = ad.mediaContent

            // These assets are not guaranteed to be present. Check that they are before
            // showing or hiding them.
            (view.bodyView as? UILabel)?.text = ad.body
            view.bodyView?.isHidden = ad.body == nil

            (view.callToActionView as? UIButton)?.setTitle(ad.callToAction, for: UIControl.State.normal)
            view.callToActionView?.isHidden = ad.callToAction == nil

            (view.iconView as? UIImageView)?.image = ad.icon?.image
            view.iconView?.isHidden = ad.icon == nil

            (view.starRatingView as? UIImageView)?.image = self.imageForStars(ad.starRating?.floatValue ?? 0)
            view.starRatingView?.isHidden = ad.starRating == nil

            (view.storeView as? UILabel)?.text = ad.store
            view.storeView?.isHidden = ad.store == nil

            (view.priceView as? UILabel)?.text = ad.price
            view.priceView?.isHidden = ad.price == nil

            (view.advertiserView as? UILabel)?.text = ad.advertiser
            view.advertiserView?.isHidden = ad.advertiser == nil

            // In order for the SDK to process touch events properly, user interaction
            // should be disabled.
            view.callToActionView?.isUserInteractionEnabled = false

            // Set ourselves as the ad delegate to be notified of native ad events.
            ad.delegate = self

            view.nativeAd = ad

            self._isLoaded = true
            self._bridge.callCpp(self._messageHelper.onLoaded)
        }
    }

    func adLoader(_ adLoader: GADAdLoader, didFailToReceiveAdWithError error: GADRequestError) {
        Thread.runOnMainThread {
            self._logger.debug("\(kTag): \(#function): \(error.localizedDescription)")
            self._bridge.callCpp(self._messageHelper.onFailedToLoad, EEJsonUtils.convertDictionary(toString: [
                "code": (error as NSError).code,
                "message": error.localizedDescription
            ]))
        }
    }

    func adLoaderDidFinishLoading(_ adLoader: GADAdLoader) {
        Thread.runOnMainThread {
            self._logger.debug("\(kTag): \(#function)")
        }
    }

    func nativeAdWillPresentScreen(_ nativeAd: GADUnifiedNativeAd) {
        Thread.runOnMainThread {
            self._logger.debug("\(kTag): \(#function)")
        }
    }

    func nativeAdDidRecordImpression(_ nativeAd: GADUnifiedNativeAd) {
        Thread.runOnMainThread {
            self._logger.debug("\(kTag): \(#function)")
        }
    }

    func nativeAdDidRecordClick(_ nativeAd: GADUnifiedNativeAd) {
        Thread.runOnMainThread {
            self._logger.debug("\(kTag): \(#function)")
        }
    }

    func nativeAdWillLeaveApplication(_ nativeAd: GADUnifiedNativeAd) {
        Thread.runOnMainThread {
            self._logger.debug("\(kTag): \(#function)")
            self._bridge.callCpp(self._messageHelper.onClicked)
        }
    }

    func nativeAdWillDismissScreen(_ nativeAd: GADUnifiedNativeAd) {
        Thread.runOnMainThread {
            self._logger.debug("\(kTag): \(#function)")
        }
    }

    func nativeAdDidDismissScreen(_ nativeAd: GADUnifiedNativeAd) {
        Thread.runOnMainThread {
            self._logger.debug("\(kTag): \(#function)")
        }
    }
}

 */
